brave/brave-core
# Add support for Leo premium

Resolves https://github.com/brave/brave-browser/issues/32738
Resolves https://github.com/brave/brave-browser/issues/33143
Serverside PR https://github.com/brave/aichat/pull/36
Security review https://github.com/brave/reviews/issues/1424

Integrates with the SKUs SDK to:
1. Expose an API `GetPremiumStatus` to frontends that returns whether there are valid premium credentials available.  Uses [CredentialSummary](https://github.com/brave/brave-core/blob/3da0ac5ffefe0e34306450b521afb85ec78f7b27/components/skus/common/skus_sdk.mojom#L12C3-L12C20) from the SKUs SDK to determine if there are credentials available.
2. Fetch credentials to be included in the Cookie header in http requests to our premium AI chat endpoint.  These credentials will be validated in our server code https://github.com/brave/aichat/pull/36. Uses [PrepareCredentialsPreparation](https://github.com/brave/brave-core/blob/3da0ac5ffefe0e34306450b521afb85ec78f7b27/components/skus/common/skus_sdk.mojom#L11) from the SKUs SDK.

Most of this logic has is handled by a new AIChatCredentialManager class.

Because the call to the SKUs sdk `PrepareCredentialsPresentation` actually consumes the credential, we must be careful to save credentials retrieved from the SDK but not successfully used in an http request.  If the server is down or the user has network issues, they could use all of them without actually making a successful call.  To handle this, I added logic that checks for a valid credential stored in local state prefs before making a call to the SKU SDK, and puts credentials back into the cache upon server error or request timeout.



## Submitter Checklist:

  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:
1. Create an account on account.bravesoftware.com (creds in 1password), purchase leo subscription. You can use 424242... (repeating) for the Stripe credit card
2. Start the browser using the dev / staging environments (some parts of this run on staging, some on dev) `npm run start -- --enable-features=AIChat --env-leo=staging --env-ai-chat.bsg=dev --env-ai-chat-premium.bsg=dev`
3. If that part of the integration worked, you should see your credentials in brave://skus-internals, and you will use those credentials for API access by default
4. To verify the premium credentials are actually being used, you may need to add some logging
